stdlib/public/Reflection/TypeRefBuilder.cpp

```diff
@@ -263,23 +309,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
   if (Found != FieldTypeInfoCache.end())
     return Found->second;
 
+  if (auto FD = getFieldDescriptorFromProvider(*MangledName))
+    return *FD;
+
   // Heuristic: find the outermost Module node available, and try to parse the
   // ReflectionInfos with a matching name first.
   auto ModuleName = FindOutermostModuleName(Node);
   // If we couldn't find a module name or the type is imported (__C module) we
   // don't any useful information on which image to look for the type.
   if (ModuleName && ModuleName != llvm::StringRef("__C"))
     for (size_t i = 0; i < ReflectionInfos.size(); ++i)
-      if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
-                             ModuleName))
-        if (auto FD = findFieldDescriptorAtIndex(i, *MangledName))
-          return *FD;
+      if ((!ExternalProvider ||
+           ExternalProvider->shouldParseReflectionInfoWithId(i)))
+        if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
+                               ModuleName))
+          if (auto FD =
+                  findFieldDescriptorAtIndex(i, *MangledName))
+            return *FD;
 
   // On failure, fill out the cache, ReflectionInfo by ReflectionInfo,
   // until we find the field descriptor we're looking for.
   for (size_t i = 0; i < ReflectionInfos.size(); ++i)
-    if (auto FD = findFieldDescriptorAtIndex(i, *MangledName))
-      return *FD;
+    if ((!ExternalProvider ||
+         ExternalProvider->shouldParseReflectionInfoWithId(i)))
```

Comment:
would
```
if (ExternalProvider && !ExternalProvider->shouldParseReflectionInfoWithId(i)))
  continuel
```
be easier to read?

Reply:
"shouldParse" doesn't explain why. Should we name it "didCacheReflectionInfo(i)" or simething more descriptive?

Reply:
I named the ExternalProvider in a deliberately not specific way just in case someone else wanted to implement the interface with a different strategy. I can change it to be more concrete though.